accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+
+### Extension Member Resolution
+
+The declaration introduces an extension. The extension's `on` type defines which types are being extended.
+
+For any member access, `x.foo`, `x.bar()`, `x.baz = 42`, `x(42)` or `x + y`, the language first checks whether the static type of `x` has a member with the same base name as the operation. That is, if it has a corresponding instance member, respectively, a `foo` method or getter or a `foo=` setter. a `bar` member or `bar=` setter, a `baz` member or `baz=` setter, a `call` method, or a `+` method. If so, then the operation is unaffected by extensions. *This check does not care whether the invocation is otherwise correct, based on number or type of the arguments, it only checks if there is a member at all.*
+
+(The types `dynamic` and `Never` are considered as having all members, the type `void` is always a compile-time error when used in a receiver position, so none of these can ever be affected by static extension methods).
+
+If there is no such member, the operation is currently a compile-time error. In that case, all extensions in scope are checked for whether they apply. An extension applies to a member access if the static type of the receiver is a subtype of the `on` type of the extension *and* the extension has an instance member with the same base name as the operation. 
+
+For generic extensions, standard type inference is used to infer the type arguments. As an example, take the following extension:
+
+```dart
+extension MyList<T extends Comparable<T>> on List<T> {
+  void quickSort() { ... }
+}
+```
+
+and a member access like:
+
+```dart
+Uint8List bytes = ...;
+bytes.quickSort();
+```
+
+Here we perform type inference equivalent to what we would do for:
+
+```dart
+class MyList<T extends Comparable<T>> {
+  MyList(List<T> argument);
+  void quickSort() { ... }
+}
+...
+Uint8List bytes = ...;
+... MyList(bytes).quickSort() ... // Infer type argument of MyList here.
+```
+
+This is inference based on static types only. The inferred type argument becomes the value of `T` for the function invocation that follows.
+
+If the inference fails, or if the synthetic constructor invocation would not be valid for any other reason, then the extension does not apply. If the extension does not have an instance member with the base name `quickSort`, it does not apply.
```

Comment:
"If an extension does not have..."?  Otherwise, I'm not sure what extension we're talking about.